pkg/serviceaccounts/controllers/create_dockercfg_secrets.go

```diff
@@ -0,0 +1,312 @@
+		glog.Error(err)
+	}
+}
+
+// serviceAccountUpdated reacts to a ServiceAccount update (or re-list) by ensuring a corresponding ServiceAccountToken Secret exists
+func (e *DockercfgController) serviceAccountUpdated(oldObj interface{}, newObj interface{}) {
+	newServiceAccount := newObj.(*api.ServiceAccount)
+
+	if err := e.createDockercfgSecretIfNeeded(newServiceAccount); err != nil {
+		glog.Error(err)
+	}
+}
+
+func getDockercfgSecretNamePrefix(serviceAccount *api.ServiceAccount) string {
+	return serviceAccount.Name + "-dockercfg-"
+}
+func getTokenSecretNamePrefix(serviceAccount *api.ServiceAccount) string {
+	return serviceAccount.Name + "-token-"
+}
+
+// createDockercfgSecretIfNeeded makes sure at least one ServiceAccountToken secret exists, and is included in the serviceAccount's Secrets list
+func (e *DockercfgController) createDockercfgSecretIfNeeded(serviceAccount *api.ServiceAccount) error {
+
+	// look for an ImagePullSecret in the form
+	dockercfgSecretName := ""
+	foundDockercfgImagePullSecret := false
+	for _, pullSecret := range serviceAccount.ImagePullSecrets {
+		if strings.HasPrefix(pullSecret.Name, getDockercfgSecretNamePrefix(serviceAccount)) {
+			foundDockercfgImagePullSecret = true
+			dockercfgSecretName = pullSecret.Name
+			break
+		}
+	}
+	foundDockercfgMountableSecret := false
+	for _, mountableSecret := range serviceAccount.Secrets {
+		if strings.HasPrefix(mountableSecret.Name, getDockercfgSecretNamePrefix(serviceAccount)) {
+			foundDockercfgMountableSecret = true
+			dockercfgSecretName = mountableSecret.Name
+			break
+		}
+	}
+
+	switch {
+	// if we already have a docker pull secret, simply return
+	case foundDockercfgImagePullSecret && foundDockercfgMountableSecret:
+		return nil
+
+	case foundDockercfgImagePullSecret && !foundDockercfgMountableSecret, !foundDockercfgImagePullSecret && foundDockercfgMountableSecret:
+		staleDecision, err := e.createDockerPullSecretReference(serviceAccount, dockercfgSecretName)
+		if staleDecision || kapierrors.IsConflict(err) {
+			// nothing to do.  Our choice was stale or we got a conflict.  Either way that means that the service account was updated.  We simply need to return because we'll get an update notification later
+			return nil
+		}
+
+		return err
+
+	}
+
+	// if we get here, then we need to create a new dockercfg secret
+	dockercfgSecret, err := e.createDockerPullSecret(serviceAccount)
+	if err != nil {
+		return err
+	}
+
+	staleDecision, err := e.createDockerPullSecretReference(serviceAccount, dockercfgSecret.Name)
+	if staleDecision || kapierrors.IsConflict(err) {
+		// nothing to do.  Our choice was stale or we got a conflict.  Either way that means that the service account was updated.  We simply need to return because we'll get an update notification later
+		// we do need to clean up our dockercfgSecret
+		if err := e.client.Secrets(dockercfgSecret.Namespace).Delete(dockercfgSecret.Name); err != nil {
```

Comment:
does the token need deleting as well, or is that handled by another controller? worth a comment if so